File: mycuration/src/main/java/com/phicdy/mycuration/ui/RegisterFilterActivity.java
package com.phicdy.mycuration.ui;

import android.content.Context;
import android.os.Bundle;
import android.support.v7.app.ActionBarActivity;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;

import com.phicdy.mycuration.R;
import com.phicdy.mycuration.db.DatabaseAdapter;
import com.phicdy.mycuration.filter.Filter;
import com.phicdy.mycuration.rss.Feed;
import com.phicdy.mycuration.tracker.GATrackerHelper;

import java.util.ArrayList;

import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper;

public class RegisterFilterActivity extends ActionBarActivity {

	private DatabaseAdapter dbAdapter;
	private String[] feedTitles;
	private ArrayList<Feed> feedsList;
	private int selectedFeedId;
	private int editFilterId;

	EditText etTitle;
	EditText etKeyword;
	EditText etFilterUrl;
	private Spinner targetFeedSpin;

	private static final int NEW_FILTER_ID = -1;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_register_filter);
		
		dbAdapter = DatabaseAdapter.getInstance(this);
		editFilterId = getIntent().getIntExtra(FilterListFragment.KEY_EDIT_FILTER_ID, NEW_FILTER_ID);
		initView();
		initData(editFilterId);

		GATrackerHelper.sendScreen(getTitle().toString());
	}
	
	private void initView() {
		setTitle(R.string.add_filter);

		etKeyword = (EditText) findViewById(R.id.filterKeyword);
		etFilterUrl = (EditText) findViewById(R.id.filterUrl);
		etTitle = (EditText) findViewById(R.id.filterTitle);

		//Set spinner
		targetFeedSpin = (Spinner)findViewById(R.id.targetFeed);
		targetFeedSpin.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
			@Override
			public void onItemSelected(AdapterView<?> parent, View view,
					int position, long id) {
				//Update selected feed ID
				selectedFeedId = feedsList.get(position).getId();
			}
	
			@Override
			public void onNothingSelected(AdapterView<?> arg0) {
			}
			
		});
		
		//If register button clicked, insert filter into DB
		Button registerButton = (Button)findViewById(R.id.registerFilter);
		registerButton.setOnClickListener(new View.OnClickListener() {

			@Override
			public void onClick(View v) {
				String keywordText = etKeyword.getText().toString();
				String filterUrlText = etFilterUrl.getText().toString();
				String titleText = etTitle.getText().toString();

				//Check title and etKeyword or filter URL has the text
				if (titleText.equals("")) {
					Toast.makeText(RegisterFilterActivity.this, R.string.title_empty_error, Toast.LENGTH_SHORT).show();
				} else if ((keywordText.equals("")) && (filterUrlText.equals(""))) {
					Toast.makeText(RegisterFilterActivity.this, R.string.both_keyword_and_url_empty_error, Toast.LENGTH_SHORT).show();
				} else if (keywordText.equals("%") || filterUrlText.equals("%")) {
					Toast.makeText(RegisterFilterActivity.this, R.string.percent_only_error, Toast.LENGTH_SHORT).show();
				} else {
					boolean result = false;
					if (editFilterId == NEW_FILTER_ID) {
						// Add new filter
						result = dbAdapter.saveNewFilter(titleText, selectedFeedId, keywordText, filterUrlText);
					} else {
						// Edit
						result = dbAdapter.updateFilter(editFilterId, titleText, keywordText, filterUrlText, selectedFeedId);
					}
					if (result) {
						Toast.makeText(getApplicationContext(), getString(R.string.filter_saved), Toast.LENGTH_SHORT).show();
					}else {
						Toast.makeText(getApplicationContext(), getString(R.string.filter_saved_error), Toast.LENGTH_SHORT).show();
					}
					finish();
				}
			}
		});
		
		//If cancel button clicked, back to filters list 
		Button cancelButton = (Button)findViewById(R.id.cancelRegisterFilter);
		cancelButton.setOnClickListener(new View.OnClickListener() {

			@Override
			public void onClick(View v) {
				finish();
			}
		});
	}

	@Override
	protected void attachBaseContext(Context newBase) {
		super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase));
	}
	
	private void initData(int editFilterId) {
		feedsList = dbAdapter.getAllFeedsWithoutNumOfUnreadArticles();
		if(feedsList.size() == 0) {
			finish();
		}else {
			// Init feed title list for spinner
			feedTitles = new String[feedsList.size()];
			for (int i = 0; i < feedsList.size(); i++) {
				feedTitles[i] = feedsList.get(i).getTitle();
			}
		}
		//Set default selected Feed ID
		selectedFeedId = feedsList.get(0).getId();

		// Set data for spinner
		ArrayAdapter<String> aa = new ArrayAdapter<String>(RegisterFilterActivity.this,android.R.layout.simple_spinner_item, feedTitles);
		aa.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
		targetFeedSpin.setAdapter(aa);

		// Edit
		if (editFilterId != NEW_FILTER_ID) {
			Filter editFilter = dbAdapter.getFilterById(editFilterId);
			if (editFilter != null) {
				etTitle.setText(editFilter.getTitle());
				etFilterUrl.setText(editFilter.getUrl());
				etKeyword.setText(editFilter.getKeyword());
				for (int i = 0; i < feedTitles.length; i++) {
					if (feedsList.get(i).getId() == editFilter.getFeedId()) {
						targetFeedSpin.setSelection(i);
						selectedFeedId = i;
						break;
					}
				}
			}
		}

	}
}

File: feature_register_filter/src/main/java/com/phicdy/feature_register_filter/RegisterFilterActivity.kt
package com.phicdy.feature_register_filter

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.lifecycleScope
import com.google.android.material.textfield.TextInputEditText
import com.phicdy.mycuration.entity.Feed
import com.phicdy.mycuration.feature.util.changeTheme
import com.phicdy.mycuration.tracker.TrackerHelper
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.components.ActivityComponent
import dagger.hilt.android.qualifiers.ActivityContext
import dagger.hilt.android.scopes.ActivityScoped
import kotlinx.coroutines.launch
import javax.inject.Inject


@AndroidEntryPoint
class RegisterFilterActivity : AppCompatActivity(), RegisterFilterView {

    @Inject
    lateinit var presenter: RegisterFilterPresenter

    private lateinit var etTitle: TextInputEditText
    private lateinit var etKeyword: TextInputEditText
    private lateinit var etFilterUrl: TextInputEditText
    private lateinit var tvTargetRss: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register_filter)

        initView()
        lifecycleScope.launch {
            presenter.create()
        }
    }

    private fun initView() {
        etKeyword = findViewById(R.id.filterKeyword)
        etFilterUrl = findViewById(R.id.filterUrl)
        etTitle = findViewById(R.id.filterTitle)

        //Set spinner
        tvTargetRss = findViewById(R.id.tv_target_rss)
        tvTargetRss.setOnClickListener {
            val intent = Intent(this@RegisterFilterActivity, SelectFilterTargetRssActivity::class.java)
            val bundle = Bundle()
            bundle.putParcelableArrayList(SelectFilterTargetRssActivity.TARGET_RSS, presenter.selectedFeedList())
            intent.putExtras(bundle)
            startActivityForResult(intent, TARGET_FEED_SELECT_REQUEST)
        }

        val toolbar = findViewById<Toolbar>(R.id.toolbar_register_filter)
        setSupportActionBar(toolbar)
        supportActionBar?.title = getString(R.string.add_filter)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setDisplayShowHomeEnabled(true)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode != Activity.RESULT_OK) return
        if (requestCode == TARGET_FEED_SELECT_REQUEST) {
            data?.extras?.let {
                val list = it.getParcelableArrayList<Feed>(KEY_SELECTED_FEED) ?: arrayListOf()
                presenter.setSelectedFeedList(list)
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_register_filter, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            // For arrow button on toolbar
            android.R.id.home -> finish()
            else -> lifecycleScope.launchWhenStarted { presenter.optionItemClicked(item) }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        changeTheme()
    }

    override fun filterKeyword(): String {
        return etKeyword.text.toString()
    }

    override fun filterUrl(): String {
        return etFilterUrl.text.toString()
    }

    override fun filterTitle(): String {
        return etTitle.text.toString()
    }

    override fun setFilterTitle(title: String) {
        etTitle.setText(title)
    }

    override fun setFilterTargetRss(rss: String) {
        tvTargetRss.text = rss
    }

    override fun setMultipleFilterTargetRss() {
        tvTargetRss.setText(R.string.multiple_target_rss)
    }

    override fun resetFilterTargetRss() {
        tvTargetRss.setText(R.string.target_rss)
    }

    override fun setFilterUrl(url: String) {
        etFilterUrl.setText(url)
    }

    override fun setFilterKeyword(keyword: String) {
        etKeyword.setText(keyword)
    }

    override fun handleEmptyTitle() {
        Toast.makeText(this@RegisterFilterActivity, R.string.title_empty_error, Toast.LENGTH_SHORT).show()
        TrackerHelper.sendButtonEvent(getString(R.string.add_new_filter_no_title))
    }

    override fun handleEmptyCondition() {
        Toast.makeText(this@RegisterFilterActivity, R.string.both_keyword_and_url_empty_error, Toast.LENGTH_SHORT).show()
        TrackerHelper.sendButtonEvent(getString(R.string.add_new_filter_no_keyword_url))
    }

    override fun handlePercentOnly() {
        Toast.makeText(this@RegisterFilterActivity, R.string.percent_only_error, Toast.LENGTH_SHORT).show()
    }

    override fun showSaveSuccessToast() {
        Toast.makeText(applicationContext, getString(R.string.filter_saved), Toast.LENGTH_SHORT).show()
    }

    override fun showSaveErrorToast() {
        Toast.makeText(
            applicationContext,
            getString(R.string.filter_saved_error),
            Toast.LENGTH_SHORT
        ).show()
    }

    override fun trackEdit() {
        TrackerHelper.sendButtonEvent(getString(R.string.update_filter))
    }

    override fun trackRegister() {
        TrackerHelper.sendButtonEvent(getString(R.string.add_new_filter))
    }

    override fun handleEmptyFeed() {
        Toast.makeText(
            applicationContext,
            getString(R.string.target_rss_empty_error),
            Toast.LENGTH_SHORT
        ).show()
    }

    companion object {
        const val KEY_EDIT_FILTER_ID = "editFilterId"
        const val KEY_SELECTED_FEED = "keySelectedFeed"
        private const val NEW_FILTER_ID = -1
        private const val TARGET_FEED_SELECT_REQUEST = 1000
    }

    @Module
    @InstallIn(ActivityComponent::class)
    object RegisterFilterModule {
        @ActivityScoped
        @Provides
        fun provideEditFilterId(@ActivityContext activity: Context): Int =
            (activity as AppCompatActivity).intent.getIntExtra(KEY_EDIT_FILTER_ID, NEW_FILTER_ID)

        @ActivityScoped
        @Provides
        fun provideRegisterFilterView(@ActivityContext activity: Context): RegisterFilterView =
                activity as RegisterFilterView
    }
}
